pkg/controller/job/job_controller.go

```diff
@@ -1871,7 +1871,7 @@ func countReadyPods(pods []*v1.Pod) int32 {
 // PodReplacementPolicy controls when we recreate pods if they are marked as terminating
 // Failed means that we recreate only once the pod has terminated.
 func onlyReplaceFailedPods(job *batch.Job) bool {
-	if feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) && *job.Spec.PodReplacementPolicy == batch.Failed {
+	if feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) && job.Spec.PodReplacementPolicy != nil && *job.Spec.PodReplacementPolicy == batch.Failed {
```

Comment:
Oh I see, the issue is when `JobPodReplacementPolicy` is disabled on API server (so no defaulting), but enabled in controller-manager. If this is the case, I think there should be a comment to explain why the check is needed. 

Wondering if this scenario could occur in production when the versions of kube-apiserver and kube-controller-manager are skewed, during upgrades or downgrades.

Reply:
Updated with a comment.  Interestingly enough we had a case for this in integration but it didn't seem to hit.

Reply:
What do you mean by "it didn't seem to hit"?

Reply:
We had a case where we should get a set fault in integration but I can only see it happen during feature enablement/disablement.

I can reproduce the failure in the unit tests though.

Reply:
I see, I think a unit test on `syncJob` level should be enough in that case.

Reply:
prefer to use `ptr.Deref`, which accepts a default value.

Reply:
nvm, this is good enough
/hold cancel